x-pack/lib/helpers/elasticsearch_options.rb

```diff
@@ -146,20 +155,58 @@ def extract_module_settings(settings)
 
     private
 
-    def check_cloud_id_configuration!(feature, settings, prefix)
-      return if !settings.set?("#{prefix}#{feature}.elasticsearch.hosts")
+    def validate_authentication!(feature, settings, prefix)
+      cloud_id = settings.set?("#{prefix}#{feature}.elasticsearch.cloud_id")
+      hosts = settings.set?("#{prefix}#{feature}.elasticsearch.hosts")
+      cloud_auth = settings.set?("#{prefix}#{feature}.elasticsearch.cloud_auth")
+      api_key = settings.set?("#{prefix}#{feature}.elasticsearch.api_key")
+      username = settings.set?("#{prefix}#{feature}.elasticsearch.username")
+      password = settings.set?("#{prefix}#{feature}.elasticsearch.password")
+      basic_auth = username && password
+
+      if cloud_auth && (username || password)
+        raise ArgumentError.new(
+          "Both #{prefix}#{feature}.elasticsearch.cloud_auth and " +
+          "#{prefix}#{feature}.elasticsearch.username/password " +
+          "specified, please only use one of those"
+        )
+      end
 
-      raise ArgumentError.new("Both \"#{prefix}#{feature}.elasticsearch.cloud_id\" and " +
-                              "\"#{prefix}#{feature}.elasticsearch.hosts\" specified, please only use one of those.")
-    end
+      if username && !password
+        raise(ArgumentError,
+          "When using #{prefix}#{feature}.elasticsearch.username, " +
+          "#{prefix}#{feature}.elasticsearch.password must also be set"
+        )
+      end
 
-    def check_cloud_auth_configuration!(feature, settings, prefix)
-      return if !settings.set?("#{prefix}#{feature}.elasticsearch.username") &&
-                !settings.set?("#{prefix}#{feature}.elasticsearch.password")
+      if cloud_id
+        if hosts
+          raise(ArgumentError,
+            "Both #{prefix}#{feature}.elasticsearch.cloud_id and " +
+            "#{prefix}#{feature}.elasticsearch.hosts specified, please only use one of those"
+          )
+        end
+
+        unless (cloud_auth || basic_auth)
```

Comment:
When a user specifies `cloud_id` and `api_key`, this fails. Is that desired?

Reply:
good catch ... we should in fact be able to specify any auth with cloud_id.